src/main/java/dan200/computercraft/core/computer/ComputerThread.java

```diff
@@ -6,227 +6,304 @@
+     */
+    private static final class TaskExecutor implements Runnable
     {
-        Object queueObject = computer;
-        
-        if (queueObject == null)
+        private TaskRunner runner;
+        private Thread thread;
+
+        @Override
+        public void run()
         {
-            queueObject = m_defaultQueue;
+            try
+            {
+                while( true )
+                {
+                    // Wait for an active queue to execute
+                    BlockingQueue<ITask> queue = s_computerTasksActive.take();
+
+                    // If threads should be stopped then return
+                    synchronized( s_stateLock )
+                    {
+                        if( s_stopped ) return;
+                    }
+
+                    execute( queue );
+                }
+            }
+            catch( InterruptedException ignored )
+            {
+            }
         }
-        
-        LinkedBlockingQueue<ITask> queue = m_computerTasks.get(queueObject);
 
-        if (queue == null)
+        private void execute( BlockingQueue<ITask> queue )
         {
-            m_computerTasks.put(queueObject, queue = new LinkedBlockingQueue<ITask>(256));
+            ITask task = queue.remove();
+
+            if( thread == null || !thread.isAlive() )
+            {
+                runner = new TaskRunner();
+
+                SecurityManager manager = System.getSecurityManager();
+                final ThreadGroup group = manager == null ? Thread.currentThread().getThreadGroup() : manager.getThreadGroup();
+                Thread thread = this.thread = new Thread( group, runner, "ComputerCraft-Computer-Runner" + s_counter.getAndIncrement() );
+                thread.setDaemon( true );
+                thread.start();
+            }
+
+            // Execute the task
+            runner.submit( task );
+
+            try
+            {
+                // If we timed out rather than exiting:
+                boolean done = runner.await( 7000 );
+                if( !done )
+                {
+                    // Attempt to soft then hard abort
+                    Computer computer = task.getOwner();
+                    if( computer != null )
+                    {
+                        computer.abort( false );
+
+                        done = runner.await( 1500 );
+                        if( !done )
+                        {
+                            computer.abort( true );
+                            done = runner.await( 1500 );
+                        }
+                    }
+
+                    // Interrupt the thread
+                    if( !done )
+                    {
+                        thread.interrupt();
+                        thread = null;
+                        runner = null;
+                    }
+                }
+            }
+            catch( InterruptedException ignored )
+            {
```

Comment:
Reset the interrupt flag??
```java
Thread.currentThread().interrupt();
```
http://www.yegor256.com/2015/10/20/interrupted-exception.html